evm-gateway/src/EVMGateway.ts

```diff
@@ -171,23 +166,30 @@ export class EVMGateway<T extends ProvableBlock> {
     const commandWord = getBytes(command);
     const flags = commandWord[0];
     const isDynamic = (flags & 0x01) != 0;
-
-    let slot = toBigInt(
-      await this.executeOperation(commandWord[1], constants, requests)
-    );
-
-    // If there are multiple path elements, recursively hash them solidity-style to get the final slot.
-    for (let j = 2; j < 32 && commandWord[j] != 0xff; j++) {
-      const index = await this.executeOperation(
-        commandWord[j],
-        constants,
-        requests
-      );
-      slot = toBigInt(
-        solidityPackedKeccak256(['bytes', 'uint256'], [index, slot])
-      );
+    let slot = 0n;
+    for (let j = 1; j < 32; j++) {
+      const op = commandWord[j];
+      if (op === OP_END) break;
+      const opcode = op & 0xe0; // upper 3
+      const operand = op & 0x1f; // lower 5
+      switch (opcode) {
+        case OP_FOLLOW_CONST: {
+          slot = followSolidityMapping(slot, constants[operand]);
+          break;
+        }
+        case OP_FOLLOW_REF: {
+          const storage = await requests[operand];
+          slot = followSolidityMapping(slot, await storage.value());
+          break;
+        }
+        case OP_ADD_CONST: {
+          slot += uint256FromBytes(constants[operand]);
+          break;
+        }
+        default:
+          throw new Error(`Unrecognized opcode: ${opcode}`);
+      }
```

Comment:
I do think it was clearer when there was a function that executes a single opcode. What's the motivation for moving it into the loop?

Reply:
It certainly could if you prefer that approach.

To me, it's clearer when you can read all of the logic together, as that function incurs more boilerplate passing state around than the logic itself.  Using the v2 branch as guidance, almost every op is a trivial one-liner.

Instead, I abstracted away the path-following code into `followSolidityMapping()` since that was duplicated in the original code.

The underlying structural change is that the initial slot offset (old L#174) is no longer special, since in this PR, that is equivalent to just using OP_ADD to the first instruction, so the loop starts at `j = 1` and `executeOperation()` only had one callsite.

Reply:
Fair enough. Let's leave it as-is, then.